### PR TITLE
feat(plasma-web): navigate by Shift + Arrow Left\Right\Up\Down

### PR DESCRIPTION
**Pull request by Yakutoc**
Для навигации только по доступным дням необходимо использовать комбинацию клавиш Shift + Arrow  Left\Right\Up\Down

Было: 

- Клавиатурная навигация игнорировала **disabled date**, а так же ограничения по мин/макс значениям. 

Стало: 

- Навигация с помощью комбинации клавиш Shift + Arrow  Left\Right\Up\Down идет **только** по доступным дням
- Клавиатурная навигация не может выйти за пределы ограничения по мин/макс значениям. 

--------------------------

:baby_chick: Download canary assets:

[colors--canary.273.4055599429.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.273.4055599429.xml)  
[color_metro_ios-swift--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_ios-swift--canary.273.4055599429.swift)  
[color_metro_kotlin--canary.273.4055599429.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_kotlin--canary.273.4055599429.kt)  
[color_metro_react-native--canary.273.4055599429.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_react-native--canary.273.4055599429.ts)  
[color_sbermarket_ios-swift--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.273.4055599429.swift)  
[color_sbermarket_kotlin--canary.273.4055599429.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.273.4055599429.kt)  
[color_sbermarket_react-native--canary.273.4055599429.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.273.4055599429.ts)  
[color_sberprime_ios-swift--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.273.4055599429.swift)  
[color_sberprime_kotlin--canary.273.4055599429.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.273.4055599429.kt)  
[color_sberprime_react-native--canary.273.4055599429.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.273.4055599429.ts)  
[color_selgros_ios-swift--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_ios-swift--canary.273.4055599429.swift)  
[color_selgros_kotlin--canary.273.4055599429.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_kotlin--canary.273.4055599429.kt)  
[color_selgros_react-native--canary.273.4055599429.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_react-native--canary.273.4055599429.ts)  
[color_smbusiness_ios-swift--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_ios-swift--canary.273.4055599429.swift)  
[color_smbusiness_kotlin--canary.273.4055599429.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_kotlin--canary.273.4055599429.kt)  
[color_smbusiness_react-native--canary.273.4055599429.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_react-native--canary.273.4055599429.ts)  
[PlasmaTokensColor--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.273.4055599429.swift)  
[typo_mage_ios-swift--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.273.4055599429.swift)  
[typo_mage_kotlin--canary.273.4055599429.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.273.4055599429.kt)  
[typo_mage_react-native--canary.273.4055599429.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.273.4055599429.ts)  
[typo_plasma_ios-swift--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.273.4055599429.swift)  
[typo_plasma_kotlin--canary.273.4055599429.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.273.4055599429.kt)  
[typo_plasma_react-native--canary.273.4055599429.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.273.4055599429.ts)  
[typo_ruler_ios-swift--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.273.4055599429.swift)  
[typo_ruler_kotlin--canary.273.4055599429.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.273.4055599429.kt)  
[typo_ruler_react-native--canary.273.4055599429.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.273.4055599429.ts)  
[typo_sage_ios-swift--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.273.4055599429.swift)  
[typo_sage_kotlin--canary.273.4055599429.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.273.4055599429.kt)  
[typo_sage_react-native--canary.273.4055599429.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.273.4055599429.ts)  
[typo_sbermarket_ios-swift--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.273.4055599429.swift)  
[typo_sbermarket_kotlin--canary.273.4055599429.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.273.4055599429.kt)  
[typo_sbermarket_react-native--canary.273.4055599429.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.273.4055599429.ts)  
[typo_soulmate_ios-swift--canary.273.4055599429.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.273.4055599429.swift)  
[typo_soulmate_kotlin--canary.273.4055599429.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.273.4055599429.kt)  
[typo_soulmate_react-native--canary.273.4055599429.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.273.4055599429.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.130.0-canary.273.4055599429.0
  npm install @salutejs/plasma-web@1.157.0-canary.273.4055599429.0
  # or 
  yarn add @salutejs/plasma-b2c@1.130.0-canary.273.4055599429.0
  yarn add @salutejs/plasma-web@1.157.0-canary.273.4055599429.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-273/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-273/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-273/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-273/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-273/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-273/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-273/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-273/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-273/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-273/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-273/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-273/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-273/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-273/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-273/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-273/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

File: packages/plasma-web/src/components/Calendar/utils.ts [hunk 1]
```diff
@@ -289,3 +289,18 @@ export const getCalendarType = (type: CalendarStateType) => {
             return 'месяц';
     }
 };
+
+export const createDateByDataAttrs = (item: HTMLDivElement): Date | undefined => {
```
Comment (Yeti-or) [inline packages/plasma-web/src/components/Calendar/utils.ts:293]: а это где используется?

File: packages/plasma-web/src/components/Calendar/CalendarBase.tsx [hunk 2]
```diff
@@ -28,6 +28,17 @@ const StyledCalendar = styled.div`
     height: 19.5rem;
 `;
 
+const IsOutOfRange = styled.div`
```
Comment (Yeti-or) [inline packages/plasma-web/src/components/Calendar/CalendarBase.tsx:31]: это невидимый элемент ?

File: packages/plasma-web/src/components/Calendar/CalendarBase.tsx [hunk 3]
```diff
@@ -28,6 +28,17 @@ const StyledCalendar = styled.div`
...
+    position: absolute;
+    padding: 0;
+    margin: 0;
+    height: 0;
+    width: 0;
+    border: 0;
+    overflow: hidden;
+    clip: rect(0 0 0 0);
```
Comment (Yeti-or) [inline packages/plasma-web/src/components/Calendar/CalendarBase.tsx:39]: видимо он не видим из-за этого
почему не visibility hidden ?

**Review by Yeti-or — CHANGES_REQUESTED** [review]

File: packages/plasma-web/src/components/Calendar/utils.ts [hunk 4]
```diff
@@ -289,3 +289,18 @@ export const getCalendarType = (type: CalendarStateType) => {
...
+    }
+
+    const year = parseInt(item.getAttribute('data-year') || '0', 10);
+    const month = parseInt(item.getAttribute('data-month-index') || '0', 10);
+    const day = parseInt(item.getAttribute('data-day') || '0', 10);
+
+    return new Date(year, month, day);
+};
+
+export const hasDateRestrict = (date?: Date, min?: Date, max?: Date): boolean | undefined =>
```
[the remark this answers is not in the document]
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/utils.ts]: @Yeti-or Это старое. Удалил, спасибо что обратил внимание.

File: packages/plasma-web/src/components/Calendar/utils.ts (hunk 1 above)
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/utils.ts]: @Yeti-or Старое, удалил.

File: packages/plasma-web/src/components/Calendar/CalendarBase.tsx (hunk 2 above)
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/CalendarBase.tsx:31]: @Yeti-or Да, это элемент который содержит в себе подсказку для screen-reader.

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-273/

File: packages/plasma-web/src/components/Calendar/CalendarBase.tsx (hunk 3 above)
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/CalendarBase.tsx:39]: @Yeti-or Перепроверю, я изначально сразу так сделал. Но screen-reader почему-то не прочел содержимое.

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🔴 Fail**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

File: packages/plasma-web/src/components/Calendar/CalendarBase.tsx (hunk 3 above)
Comment (Yeti-or) [reply packages/plasma-web/src/components/Calendar/CalendarBase.tsx:39]: Мы не можем поставить тупо aria-label и он будет доступен для ридера но не доступен визуально ?

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-273/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-273/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-273/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-273/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-273/temple-storybook/

File: packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts [hunk 5]
```diff
@@ -160,11 +305,52 @@ export const useKeyNavigation = ({ isDouble = false, size, onPrev, onNext }: Use
...
+                        newRowIndex = rowIndex;
+                        newColumnIndex = columnIndex;
+                    }
+
                     if (isVisible(outerRefs, newRowIndex, newColumnIndex)) {
                         break;
                     }
 
                     if (!isDouble || getDoubleCalendarSide(currentRowIndex) === 'first') {
+                        if (isDisabledArrowLeft) {
```
[the remark this answers is not in the document]
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts]: @Yeti-or Да, тест есть.

File: packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts [hunk 6]
```diff
@@ -177,11 +363,52 @@ export const useKeyNavigation = ({ isDouble = false, size, onPrev, onNext }: Use
...
+                        newRowIndex = rowIndex;
+                        newColumnIndex = columnIndex;
+                    }
+
                     if (isVisible(outerRefs, newRowIndex, newColumnIndex)) {
                         break;
                     }
 
                     if (!isDouble || getDoubleCalendarSide(currentRowIndex) === 'second') {
+                        if (isDisabledArrowRight) {
```
[the remark this answers is not in the document]
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts]: @Yeti-or Да, тест есть.

File: packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts [hunk 7]
```diff
@@ -24,11 +24,108 @@ const isOutOfBound = ([rowIndex, columnIndex]: number[], rowSize: number, column
...
+    columnSize,
+    minColumnIndex,
+}: DaysMetaDescription): number[] => {
+    let item = refs.current?.[newRowIndex]?.[newColumnIndex];
+
+    while (isAriaDisableItem(item) && newColumnIndex >= minColumnIndex) {
+        newColumnIndex--;
+
+        if (newColumnIndex < minColumnIndex && newRowIndex <= rowSize) {
+            newRowIndex--;
```
[the remark this answers is not in the document]
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts:101]: @Yeti-or Да, тест есть.

File: packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts [hunk 8]
```diff
@@ -24,11 +24,108 @@ const isOutOfBound = ([rowIndex, columnIndex]: number[], rowSize: number, column
...
+    // Смещение влево - начинаем поиск первого не-nullable элемента в первой строке
+    return refs.current[0].findIndex((item) => item) - 1;
+}
+
 const getCorrectIndexes = (
     refs: React.MutableRefObject<HTMLDivElement[][]>,
     [rowIndex, columnIndex]: number[],
     rowSize: number,
     columnSize: number,
+    withShift = false,
```
[the remark this answers is not in the document]
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts:160]: @Yeti-or хм, проверю, но скорее ты прав и можно без по-умолчанию.

File: packages/plasma-web/src/components/Calendar/hooks/useDays.ts [hunk 9]
```diff
@@ -115,8 +163,17 @@ const getDaysWithModifications = (
...
         dayItem.events = eventsMap.get(keyDay);
-        dayItem.disabled = disabledsMap.has(keyDay) || (min && min >= currentDate) || (max && max <= currentDate);
+        dayItem.disabled = disabledDaysMap.has(keyDay) || isOutOfRange;
+
+        dayItem.isDisabledArrowLeft = isDisabledArrowLeft(currentDate, min) || isOutOfRange;
+        dayItem.isDisabledArrowRight = isDisabledArrowRight(currentDate, max) || isOutOfRange;
+        dayItem.isDisabledArrowUp = isDisabledArrowUp(currentDate, min);
+        dayItem.isDisabledArrowDown = isDisabledArrowDown(currentDate, max);
+        dayItem.isDisabledNextMonth = isDisabledNextMonth(currentDate, max);
+        dayItem.isDisabledPreviousMonth = isDisabledPreviousMonth(currentDate, min);
```
Comment (neretin-trike) [inline packages/plasma-web/src/components/Calendar/hooks/useDays.ts:176]: Кажется, что можно эту логику упростить. По сути, тебе не надо столько дата атрибуттов накидывать, достаточно одного и в нем хранить логику всех этих проверок.

Имею ввиду, что у тебя будет условно "dayItem.isDisable", и в нём будет значение в зависимости от того, какой метод (isDisabledArrowLeft,Right,Up,Down,NextMonth,PrevMonth) вернул true быстрее

File: packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts [hunk 10]
```diff
@@ -24,11 +24,143 @@ const isOutOfBound = ([rowIndex, columnIndex]: number[], rowSize: number, column
 const isVisible = (refs: React.MutableRefObject<HTMLDivElement[][]>, row: number, column: number) =>
     refs.current?.[row]?.[column];
 
+const isAriaDisableItem = (item: HTMLDivElement) => item?.getAttribute('aria-disabled') === 'true';
+
+function isDisabledNavigateByAttr({
```
Comment (neretin-trike) [inline packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts:29]: *minor*

Тут бы я тоже использовал стрелочную функцию для поддержания единого стиля

File: packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts [hunk 11]
```diff
@@ -24,11 +24,143 @@ const isOutOfBound = ([rowIndex, columnIndex]: number[], rowSize: number, column
...
+        }
+    }
+
+    return [newRowIndex, newColumnIndex];
+};
+
+/**
+ * Метод для нахождения стартового индекса дня в следующем/предыдущем месяце
+ */
+function getCorrectColumnIndex({
```
Comment (neretin-trike) [inline packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts:132]: Здесь тож стрелочную нужно

File: packages/plasma-web/src/components/Calendar/CalendarBase.tsx (hunk 3 above)
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/CalendarBase.tsx:39]: @Yeti-or К сожалению, нет. Это **aria-live region** элемент  

```
     aria-atomic="true"
     role="alert"
     aria-live="assertive"
     aria-relevant="additions"
```   
и он должен быть "условно виден", если мы хотим, чтобы изменения его содержимого объявлялось и озвучивалось корректно. 

[ARIA Live Regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions)

File: packages/plasma-web/src/components/Calendar/hooks/useDays.ts (hunk 9 above)
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/hooks/useDays.ts:176]: @neretin-trike Тут немного иначе работает. 

Этими проверками я узнаю в каком направление нельзя передвигаться что бы на соотвествующие клавиши блокировать событие.         

Пример:  
**min** - апрель
**выключенные дни** в мае - с 1 по 7 число.

По **моей** логики через **(Shift + Left)** я не могу с 7 мая перейти на 6 число, потому что день выключенный. 

Так же через Shift + Up не получиться перейти на 30 апреля, потому что `isDisabledArrowUp === true` (апрель у нас min)

НО я могу с 7 мая перейти на 6 число через Left, а по твоей логики 7 число уже будет иметь `isDisable === true` потому что проверка **isDisabledArrowUp** вернула именно это.

File: packages/plasma-web/src/components/Calendar/Calendar.component-test.tsx [hunk 12]
```diff
@@ -431,6 +431,382 @@ describe('plasma-web: Calendar keyboard navigation', () => {
...
+            cy.get('body')
+                .find('[data-day="2"][data-month-index="4"]')
+                .first()
+                .should('have.attr', 'tabindex', '0')
+                .type(leftArrow);
+
+            cy.get('body').find(start).first().should('have.attr', 'tabindex', '0');
+        });
+
+        it.skip("Can't jump to dates outside of min max dates", () => {
```
Comment (Yeti-or) [inline packages/plasma-web/src/components/Calendar/Calendar.component-test.tsx:588]: ждем тестов

File: packages/plasma-web/src/components/Calendar/hooks/useDays.ts (hunk 9 above)
Comment (neretin-trike) [reply packages/plasma-web/src/components/Calendar/hooks/useDays.ts:176]: оке, я понял о чём ты. А что если попробовать вместо булевых значений хранить строку, в которой перечисляются направление по которым нельзя перейти? А затем при проверки их парсить и понимать, можешь перейти или нет? Если я правильно понял, то тебе только важно наличие или отсутствие этого значение, и тогда будет что-то вроде:

```ts
.... // здесь функция, которая возвращает направления, которые залочены и собирает это в строку
dayItem.isDisabledTo = 'arrowLeft, arrowRight, arrowDown, arrowUp';
```

а метод проверки похожий на:

```ts
function isDisabledNavigateByAttr({
    refs,
    payload,
    attr,
}: {
    refs: React.MutableRefObject<HTMLDivElement[][]>;
    payload: number[];
    arrow: 'arrowLeft' | 'arrowRight' | 'arrowUp' | 'arrowDown';
}) {
    const [previousRowIndex, previousColumnIndex] = payload;

    const disabledArrows = refs.current?.[previousRowIndex]?.[previousColumnIndex].dataset.isDisabledTo;

    return  disabledArrows.includes(arrow) === 'true';
}
```

Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/hooks/useDays.ts:177]: Я думал о чем то подобном, но через JSON. Твой вариант кажется имеет право на жизнь - сделаю.

File: packages/plasma-web/src/components/Calendar/Calendar.component-test.tsx (hunk 12 above)
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/Calendar.component-test.tsx]: @Yeti-or Тесты которые падали поправил.

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-273/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-273/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-273/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-273/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-273/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-273/temple-storybook/

**Review by Yeti-or — APPROVED** [review]

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-273/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-273/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-273/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-273/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-273/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-273/temple-storybook/

**Review by neretin-trike — APPROVED** [review]

File: packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts [hunk 13]
```diff
@@ -24,11 +24,147 @@ const isOutOfBound = ([rowIndex, columnIndex]: number[], rowSize: number, column
...
+}: {
+    refs: React.MutableRefObject<HTMLDivElement[][]>;
+    payload: number[];
+    key: KeyboardArrowKey;
+}) => {
+    const [previousRowIndex, previousColumnIndex] = payload;
+
+    const disabledArrowKey = refs.current?.[previousRowIndex]?.[previousColumnIndex]?.dataset.disabledArrowKey;
+
+    return disabledArrowKey ? disabledArrowKey.includes(key) : false;
```
Comment (neretin-trike) [inline packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts:42]: а тут точно тернарник нужен?* Почему не 
```ts
return disabledArrowKey?.includes(key)
```

Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/hooks/useKeyNavigation.ts:42]: > а тут точно тернарник нужен?* Почему не
> 
> ```ts
> return disabledArrowKey?.includes(key)
> ```

Скорее всего ты прав.

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-273/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🔴 Fail**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-273/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-273/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-273/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-273/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-273/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-273/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-273/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-273/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-273/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-273/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-273/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-273/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-273/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-273/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-273/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-273/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-273/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-273/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-273/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-273/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-273/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-273/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-273/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🔴 Fail**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-273/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-273/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-273/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-273/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-273/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-273/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-273/temple-storybook/

**Review by Yeti-or — APPROVED** [review]

**Pull request merged**